runtime/src/utils/buffer/write.rs

```diff
@@ -317,12 +312,15 @@ impl<B: Blob> Blob for Write<B> {
     }
 
     async fn sync(&self) -> Result<(), Error> {
-        let mut inner = self.inner.write().await;
-        inner.sync().await
+        let mut buffer = self.buffer.write().await;
+        buffer.flush(&self.blob).await?;
+        self.blob.sync().await
     }
 
     async fn close(self) -> Result<(), Error> {
-        let mut inner = self.inner.write().await;
-        inner.close().await
+        self.sync().await?;
+        // We use clone here to ensure we retain the close semantics of the blob provided (if
+        // called multiple times, the blob determines whether to error).
+        self.blob.clone().close().await
```

Comment:
I'm not in love with this approach. I think if you were concerned about this, you'd provide a clone when instantiating `write` rather than assuming we'll always clone.

I think this feels like "non-default" behavior and suggest just calling close.